src/Analyzer.elm

```diff
@@ -315,12 +409,92 @@ matchFunction (Node range expression) foundFunction =
         ( PrefixOperator exprOperator, NotFound (Operator operator) ) ->
             match exprOperator operator
 
-        -- already found function, or expression that doesn't match
+        ( LetExpression { declarations }, NotFound (LetWithPattern _) ) ->
+            declarations
+                |> List.filterMap letDestructuringPattern
+                |> List.concatMap ElmSyntaxHelpers.traversePattern
+                |> List.foldl matchExpressionPattern foundExpression
+
+        ( CaseExpression { cases }, NotFound (CaseWithPattern _) ) ->
+            cases
+                |> List.map Tuple.first
+                |> List.concatMap ElmSyntaxHelpers.traversePattern
+                |> List.foldl matchExpressionPattern foundExpression
+
+        ( LambdaExpression { args }, NotFound (LambdaWithPattern _) ) ->
+            args
+                |> List.concatMap ElmSyntaxHelpers.traversePattern
+                |> List.foldl matchExpressionPattern foundExpression
+
+        -- already found expression, argument pattern, or expression that doesn't match
         _ ->
-            foundFunction
+            foundExpression
+
+
+matchArgumentPattern : Node Pattern.Pattern -> FoundExpression -> FoundExpression
+matchArgumentPattern (Node range elmSyntaxPattern) foundExpression =
+    case foundExpression of
+        NotFound (ArgumentWithPattern pattern) ->
+            if matchPattern elmSyntaxPattern pattern then
+                foundAt range foundExpression
+
+            else
+                foundExpression
+
+        _ ->
+            foundExpression
+
+
+matchExpressionPattern : Node Pattern.Pattern -> FoundExpression -> FoundExpression
+matchExpressionPattern (Node range elmSyntaxPattern) foundExpression =
+    let
+        checkPattern pattern =
+            if matchPattern elmSyntaxPattern pattern then
+                foundAt range foundExpression
+
+            else
+                foundExpression
+    in
+    case foundExpression of
+        NotFound (LetWithPattern pattern) ->
+            checkPattern pattern
+
+        NotFound (CaseWithPattern pattern) ->
+            checkPattern pattern
+
+        NotFound (LambdaWithPattern pattern) ->
+            checkPattern pattern
+
+        _ ->
+            foundExpression
+
+
+matchPattern : Pattern.Pattern -> Pattern -> Bool
+matchPattern elmSyntaxPattern pattern =
+    case ( elmSyntaxPattern, pattern ) of
+        ( Pattern.RecordPattern _, Record ) ->
+            True
+
+        ( Pattern.TuplePattern _, Tuple ) ->
+            True
+
+        ( Pattern.AllPattern, Ignore ) ->
```

Comment:
Yes I know it's jarring, I first tried `All`, but I'm already using this variant in the `Find` type.
I can't rename `AllPattern` because it's an elm-syntax type.
What do you think a better name would be? I do like `Ignore` or `Wildcard` even if they don't match but I don't like `AllPattern`.

Reply:
I think I would probably use AllPattern and match the elm-syntax type. I don't have all the context though, so I'm happy to go with whatever you think is best (and this is a minor point anyway)

Reply:
My goal is to optimize usability for writing the analyzer rules, and I think `AllPattern` is not a very friendly name. I will leave it like that for now, renaming it later is trivial anyway.